### DISCUSSION TITLE
Rounded corners

### DISCUSSION THREAD

**Discussion by DOEHOONLEE** [ImageMagick/ImageMagick#4156] [Help]
Hi,
I was looking into how to get an image with rounded corners and found this

```
convert thumbnail.gif \
     \( +clone  -alpha extract \
        -draw 'fill black polygon 0,0 0,15 15,0 fill white circle 15,15 15,0' \
        \( +clone -flip \) -compose Multiply -composite \
        \( +clone -flop \) -compose Multiply -composite \
     \) -alpha off -compose CopyOpacity -composite  rounded_corners.png
```

but then I couldn't find what the values (after polygon and circle) mean.

Does anyone know where I can find their meanings?

Thanks!!

**Comment by snibgo** [comment]
For the parameters to "polygon" and "circle", see the documentation at https://imagemagick.org/script/command-line-options.php#draw

In your command, "15" is the radius in pixels of the quarter circles at the corners.

**Reply by DOEHOONLEE** [reply]
![Screen Shot 2021-09-07 at 11 23 16 AM](https://user-images.githubusercontent.com/20305442/132274555-d5e20a91-7f0e-4d92-b191-e5969774b1a8.png)

hmm..

I assume

* two points in "line" are the two endings and imageMagick connects two points to draw a line.
* for the "rectangle," one point is at top left and one at bottom right then it figures out the other two points to draw a square
* `x0,y0` is the center of the circle, and the difference between `x0,y0` && `x1,y1` is the radius of the circle

did I get them right?

but.. what are `wc, hc` in roundRectangle? Perhaps they are width and height? but.. what's the logic behind it??

also `a0,a1` in arc?

Thank you~!

**Reply by fmw42** [reply]
Arc and ellipse need the starting and ending angles along the curve to tell it how much of it to draw (full ellipse vs half or quarter, etc).  wc and hc are the radii (x and y) for the rounded corners. (I suspect they are actually the distance from the corner to the start and stop parts where the curve begins and ends on x and y.)

See https://legacy.imagemagick.org/Usage/draw/#primitives

**Comment by snibgo** [comment]
> but.. what are wc, hc in roundRectangle? Perhaps they are width and height? but.. what's the logic behind it??

See https://imagemagick.org/script/command-line-options.php#draw :
> A roundRectangle primitive takes the same corner points as a rectangle followed by the _width and height of the rounded corners_ to be removed.

So "roundRectangle" has two radii, for the x and y directions, and it draws parts of an ellipse. "circle" has a single radius, so it really is a circle. "arc" and "ellipse" draw (parts of) ellipses.

If you want elliptical corners, you can do it like this (bash syntax):
```
rx=50
ry=30

magick \
  -size ${rx}x${ry} xc:Black \
  -fill White \
  -draw "ellipse ${rx},${ry},${rx},${ry},180,360" \
  -write mpr:CORN +delete \
  in.png \
  \( +clone -alpha extract \
     -compose Darken \
     mpr:CORN \
     -gravity NorthWest -composite \
     \( mpr:CORN -flop \) \
     -gravity NorthEast -composite \
     \( mpr:CORN -flip \) \
     -gravity SouthWest -composite \
     \( mpr:CORN -flip -flop \) \
     -gravity SouthEast -composite \
     +write m.png \
  \) \
  -alpha off \
  -compose CopyOpacity -composite \
  out.png
```
As always, there are many possible methods.

**Reply by philocalyst** [reply]
Does this impact the input images transparency?
